src/data/ellpack_page.cuh

```diff
@@ -42,56 +40,68 @@ __forceinline__ __device__ int BinarySearchRow(
   return -1;
 }
 
+/** \brief Meta information about the ELLPACK matrix. */
+struct EllpackInfo {
+  /*! \brief Whether or not if the matrix is dense. */
+  bool is_dense;
+  /*! \brief Row length for ELLPack, equal to number of features. */
+  size_t row_stride;
+  /*! \brief Total number of bins, also used as the null index value, . */
+  size_t n_bins;
+  /*! \brief Minimum value for each feature. Size equals to number of features. */
+  common::Span<bst_float> min_fvalue;
+  /*! \brief Histogram cut pointers. Size equals to (number of features + 1). */
+  common::Span<uint32_t> feature_segments;
+  /*! \brief Histogram cut values. Size equals to (bins per feature * number of features). */
+  common::Span<bst_float> gidx_fvalue_map;
```

Comment:
These are not trivial when dataset is wide.  But we can consider optimization after having a working pipeline.

Reply:
Agreed. It's actually kept in memory.